File: java_projects/cleaned_final_projects_evosuite/commons-pool/src/main/java/org/apache/commons/pool2/impl/DefaultPooledObject.java

/*
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

package org.apache.commons.pool2.impl;

import org.apache.commons.pool2.PooledObject;
import org.apache.commons.pool2.PooledObjectState;
import org.apache.commons.pool2.TrackedUse;

import java.io.PrintWriter;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;

/**
 * This wrapper is used to track the additional information, such as state, for the pooled objects.
 *
 * <p>This class is intended to be thread-safe.
 *
 * @param <T> the type of object in the pool
 * @since 2.0
 */
public class DefaultPooledObject<T> {

    private final T object;
    private PooledObjectState state =
            PooledObjectState.IDLE; // @GuardedBy("this") to ensure transitions are valid
    private final Clock systemClock = Clock.systemUTC();
    private final Instant createInstant = now();

    private volatile Instant lastBorrowInstant = createInstant;
    private volatile Instant lastUseInstant = createInstant;
    private volatile Instant lastReturnInstant = createInstant;
    private volatile boolean logAbandoned;
    private volatile CallStack borrowedBy = NoOpCallStack.INSTANCE;
    private volatile CallStack usedBy = NoOpCallStack.INSTANCE;
    private volatile long borrowedCount;

    /**
     * Creates a new instance that wraps the provided object so that the pool can track the state of
     * the pooled object.
     *
     * @param object The object to wrap
     */
    public DefaultPooledObject(final T object) {
        this.object = object;
    }

    /**
     * Allocates the object.
     *
     * @return {@code true} if the original state was {@link PooledObjectState#IDLE IDLE}
     */
    public synchronized boolean allocate() {
        if (state == PooledObjectState.IDLE) {
            state = PooledObjectState.ALLOCATED;
            lastBorrowInstant = now();
            lastUseInstant = lastBorrowInstant;
            borrowedCount++;
            if (logAbandoned) {
                borrowedBy.fillInStackTrace();
            }
            return true;
        }
        if (state == PooledObjectState.EVICTION) {
            state = PooledObjectState.EVICTION_RETURN_TO_HEAD;
        }
        return false;
    }

    public int compareTo(final PooledObject<T> other) {
        final int compareTo = getLastReturnInstant().compareTo(other.getLastReturnInstant());
        if (compareTo == 0) {
            return System.identityHashCode(this) - System.identityHashCode(other);
        }
        return compareTo;
    }

    /**
     * Deallocates the object and sets it {@link PooledObjectState#IDLE IDLE} if it is currently
     * {@link PooledObjectState#ALLOCATED ALLOCATED} or {@link PooledObjectState#RETURNING
     * RETURNING}.
     *
     * @return {@code true} if the state was {@link PooledObjectState#ALLOCATED ALLOCATED} or {@link
     *     PooledObjectState#RETURNING RETURNING}.
     */
    public synchronized boolean deallocate() {
        if (state == PooledObjectState.ALLOCATED || state == PooledObjectState.RETURNING) {
            state = PooledObjectState.IDLE;
            lastReturnInstant = now();
            borrowedBy.clear();
            return true;
        }

        return false;
    }

    /**
     * Gets the number of times this object has been borrowed.
     *
     * @return The number of times this object has been borrowed.
     * @since 2.1
     */
    public long getBorrowedCount() {
        return borrowedCount;
    }

    public Instant getCreateInstant() {
        return createInstant;
    }

    public long getCreateTime() {
        return createInstant.toEpochMilli();
    }

    public Duration getIdleDuration() {
        final Duration elapsed = Duration.between(lastReturnInstant, now());
        return elapsed.isNegative() ? Duration.ZERO : elapsed;
    }

    public Duration getIdleTime() {
        return getIdleDuration();
    }

    public long getIdleTimeMillis() {
        return getIdleDuration().toMillis();
    }

    public Instant getLastBorrowInstant() {
        return lastBorrowInstant;
    }

    public long getLastBorrowTime() {
        return lastBorrowInstant.toEpochMilli();
    }

    public Instant getLastReturnInstant() {
        return lastReturnInstant;
    }

    public long getLastReturnTime() {
        return lastReturnInstant.toEpochMilli();
    }

    /**
     * Gets an estimate of the last time this object was used. If the class of the pooled object
     * implements {@link TrackedUse}, what is returned is the maximum of {@link
     * TrackedUse#getLastUsedInstant()} and {@link #getLastBorrowTime()}; otherwise this method
     * gives the same value as {@link #getLastBorrowTime()}.
     *
     * @return the last Instant this object was used.
     */
    public Instant getLastUsedInstant() {
        if (object instanceof TrackedUse) {
            return PoolImplUtils.max(((TrackedUse) object).getLastUsedInstant(), lastUseInstant);
        }
        return lastUseInstant;
    }

    /**
     * Gets an estimate of the last time this object was used. If the class of the pooled object
     * implements {@link TrackedUse}, what is returned is the maximum of {@link
     * TrackedUse#getLastUsedInstant()} and {@link #getLastBorrowTime()}; otherwise this method
     * gives the same value as {@link #getLastBorrowTime()}.
     *
     * @return the last time this object was used
     */
    public long getLastUsedTime() {
        return getLastUsedInstant().toEpochMilli();
    }

    public T getObject() {
        return object;
    }

    /**
     * Gets the state of this object.
     *
     * @return state
     */
    public synchronized PooledObjectState getState() {
        return state;
    }

    /** Sets the state to {@link PooledObjectState#INVALID INVALID}. */
    public synchronized void invalidate() {
        state = PooledObjectState.INVALID;
    }

    /** Marks the pooled object as {@link PooledObjectState#ABANDONED ABANDONED}. */
    public synchronized void markAbandoned() {
        state = PooledObjectState.ABANDONED;
    }

    /** Marks the pooled object as {@link PooledObjectState#RETURNING RETURNING}. */
    public synchronized void markReturning() {
        state = PooledObjectState.RETURNING;
    }

    /**
     * Gets the current instant of the clock.
     *
     * @return the current instant of the clock.
     */
    private Instant now() {
        return systemClock.instant();
    }

    public void printStackTrace(final PrintWriter writer) {
        boolean written = borrowedBy.printStackTrace(writer);
        written |= usedBy.printStackTrace(writer);
        if (written) {
            writer.flush();
        }
    }

    public void setLogAbandoned(final boolean logAbandoned) {
        this.logAbandoned = logAbandoned;
    }

    /**
     * Configures the stack trace generation strategy based on whether or not fully detailed stack
     * traces are required. When set to false, abandoned logs may only include caller class
     * information rather than method names, line numbers, and other normal metadata available in a
     * full stack trace.
     *
     * @param requireFullStackTrace the new configuration setting for abandoned object logging
     * @since 2.5
     */
    public synchronized boolean startEvictionTest() {
        if (state == PooledObjectState.IDLE) {
            state = PooledObjectState.EVICTION;
            return true;
        }
        return false;
    }

    public String toString() {
        final StringBuilder result = new StringBuilder();
        result.append("Object: ");
        result.append(object.toString());
        result.append(", State: ");
        synchronized (this) {
            result.append(state.toString());
        }
        return result.toString();
    }

    public void use() {
        lastUseInstant = now();
        usedBy.fillInStackTrace();
    }
}

File: data/recomposed_projects/llama-3-70b-instruct/signature/commons-pool/src/main/org/apache/commons/pool2/impl/DefaultPooledObject.py
from __future__ import annotations
import time
import re
from io import StringIO
import io
import typing
from typing import *
import datetime
import sys
from src.main.org.apache.commons.pool2.PooledObject import *
from src.main.org.apache.commons.pool2.PooledObjectState import *
from src.main.org.apache.commons.pool2.TrackedUse import *
from src.main.org.apache.commons.pool2.impl.CallStack import *
from src.main.org.apache.commons.pool2.impl.NoOpCallStack import *
from src.main.org.apache.commons.pool2.impl.PoolImplUtils import *


class DefaultPooledObject:

    __borrowedCount: int = 0

    __usedBy: CallStack = NoOpCallStack.INSTANCE
    __borrowedBy: CallStack = NoOpCallStack.INSTANCE
    __logAbandoned: bool = False

    __createInstant: datetime.datetime = self.__now()
    __systemClock: datetime.datetime = datetime.datetime.now()
    __state: PooledObjectState = PooledObjectState.IDLE
    __object: typing.Any = None

    __lastReturnInstant: datetime.datetime = __createInstant
    __lastUseInstant: datetime.datetime = __createInstant
    __lastBorrowInstant: datetime.datetime = __createInstant

    def use(self) -> None:
        self.__lastUseInstant = self.__now()
        self.__usedBy.fillInStackTrace()

    def toString(self) -> str:

        pass  # LLM could not translate this method

    def startEvictionTest(self) -> bool:
        if self.__state == PooledObjectState.IDLE:
            self.__state = PooledObjectState.EVICTION
            return True
        return False

    def setLogAbandoned(self, logAbandoned: bool) -> None:
        self.__logAbandoned = logAbandoned

    def printStackTrace(
        self, writer: typing.Union[io.TextIOWrapper, io.StringIO]
    ) -> None:
        written = self.__borrowedBy.printStackTrace(writer)
        written |= self.__usedBy.printStackTrace(writer)
        if written:
            writer.flush()

    def markReturning(self) -> None:
        self.__state = PooledObjectState.RETURNING

    def markAbandoned(self) -> None:
        self.__state = PooledObjectState.ABANDONED

    def invalidate(self) -> None:
        self.__state = PooledObjectState.INVALID

    def getState(self) -> PooledObjectState:
        return self.__state

    def getObject(self) -> typing.Any:
        return self.__object

    def getLastUsedTime(self) -> int:
        return self.getLastUsedInstant().timestamp() * 1000

    def getLastUsedInstant(self) -> datetime.datetime:
        if isinstance(self.__object, TrackedUse):
            return PoolImplUtils.max(
                self.__object.getLastUsedInstant(), self.__lastUseInstant
            )
        return self.__lastUseInstant

    def getLastReturnTime(self) -> int:
        return self.__lastReturnInstant.timestamp() * 1000

    def getLastReturnInstant(self) -> datetime.datetime:
        return self.__lastReturnInstant

    def getLastBorrowTime(self) -> int:
        return self.__lastBorrowInstant.timestamp() * 1000

    def getLastBorrowInstant(self) -> datetime.datetime:
        return self.__lastBorrowInstant

    def getIdleTimeMillis(self) -> int:
        return self.getIdleDuration().total_seconds() * 1000

    def getIdleTime(self) -> datetime.timedelta:
        return self.getIdleDuration()

    def getIdleDuration(self) -> datetime.timedelta:
        elapsed = datetime.timedelta.between(self.__lastReturnInstant, self.__now())
        return datetime.timedelta.ZERO if elapsed.is_negative() else elapsed

    def getCreateTime(self) -> int:
        return self.__createInstant.timestamp() * 1000

    def getCreateInstant(self) -> datetime.datetime:
        return self.__createInstant

    def getBorrowedCount(self) -> int:
        return self.__borrowedCount

    def deallocate(self) -> bool:
        if (
            self.__state == PooledObjectState.ALLOCATED
            or self.__state == PooledObjectState.RETURNING
        ):
            self.__state = PooledObjectState.IDLE
            self.__lastReturnInstant = self.__now()
            self.__borrowedBy.clear()
            return True
        return False

    def compareTo(self, other: PooledObject[typing.Any]) -> int:
        compareTo = self.getLastReturnInstant().compareTo(other.getLastReturnInstant())
        if compareTo == 0:
            return System.identityHashCode(self) - System.identityHashCode(other)
        return compareTo

    def allocate(self) -> bool:

        pass  # LLM could not translate this method

    def __init__(self, object: typing.Any) -> None:
        self.__object = object

    def __now(self) -> datetime.datetime:
        return self.__systemClock.now()
